### PR TITLE
Error 500 when creating a Tier with an already existing Name

### PR DESCRIPTION
**Pull request by JLSRKonk**
# Readiness checklist

- [x] I added/updated unit tests.
- [x] I added/updated integration tests.
- [x] I ensured that the PR title is good enough for the changelog.
- [x] I labeled the PR.

### CODE REVIEW

File: AdminUi/test/AdminUi.Tests.Integration/StepDefinitions/TiersStepDefinitions.cs [hunk 1]
```diff
@@ -1,39 +1,119 @@
...
+    private string _tierName;
 
     public TiersStepDefinitions(TiersApi tiersApi)
     {
         _tiersApi = tiersApi;
+        _tierName = string.Empty;
+    }
+
+    [Given(@"a Tier t")]
+    public async Task GivenAValidTier()
```
Comment (tnotheis) [inline AdminUi/test/AdminUi.Tests.Integration/StepDefinitions/TiersStepDefinitions.cs:25]:
```suggestion
    public async Task GivenATier()
```

File: AdminUi/test/AdminUi.Tests.Integration/StepDefinitions/TiersStepDefinitions.cs [hunk 2]
```diff
@@ -1,39 +1,119 @@
...
+
+        var requestConfiguration = _requestConfiguration.Clone();
+        requestConfiguration.ContentType = "application/json";
+        requestConfiguration.SetContent(createTierRequest);
+
+        var response = await _tiersApi.CreateTier(requestConfiguration);
+
+        var actualStatusCode = (int)response.StatusCode;
+        actualStatusCode.Should().Be(201);
+        _tierName = response.Content.Result!.Name;
```
Comment (tnotheis) [inline AdminUi/test/AdminUi.Tests.Integration/StepDefinitions/TiersStepDefinitions.cs:40]:
```suggestion
        _existingTier = response.Content.Result!.Name;
```

File: Modules/Devices/src/Devices.Application/Infrastructure/Persistence/Repository/ITiersRepository.cs [hunk 3]
```diff
@@ -8,6 +8,8 @@ public interface ITiersRepository
 {
     Task AddAsync(Tier tier, CancellationToken cancellationToken);
 
+    Task<bool> ExistsWithName(string tierName, CancellationToken cancellationToken);
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Application/Infrastructure/Persistence/Repository/ITiersRepository.cs:11]:
```suggestion
    Task<bool> ExistsWithName(TierName tierName, CancellationToken cancellationToken);
```

File: Modules/Devices/src/Devices.Infrastructure/Persistence/Repository/TiersRepository.cs [hunk 4]
```diff
@@ -26,6 +26,12 @@ public async Task AddAsync(Tier tier, CancellationToken cancellationToken)
         await _dbContext.SaveChangesAsync(cancellationToken);
     }
 
+    public async Task<bool> ExistsWithName(string tierName, CancellationToken cancellationToken)
+    {
+        var tier = await _tiersDbSet.FirstOrDefaultAsync(t => t.Name == tierName, cancellationToken);
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Infrastructure/Persistence/Repository/TiersRepository.cs:31]: Always try to transfer as little data as possible to and from the database.

```suggestion
        return await _tiersDbSet.AnyAsync(t => t.Name == tierName, cancellationToken);
```

File: Modules/Devices/test/Devices.Application.Tests/Tests/Tiers/Commands/CreateTier/HandlerTests.cs [hunk 5]
```diff
@@ -42,4 +44,18 @@ public async void Creates_a_tier_when_properties_are_valid()
         result.Id.Should().NotBeNull();
         result.Name.Should().Be(expectedTierName.Value);
     }
+
+    [Fact]
+    public async void Fails_to_create_a_tier_when_tier_name_already_exists()
+    {
+        // Arrange
+        var expectedTierName = TierName.Create("my-tier-name");
```
Comment (tnotheis) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Tiers/Commands/CreateTier/HandlerTests.cs:52]: Why/where is this tier name "expected"?

File: Modules/Devices/src/Devices.Application/Tiers/Commands/CreateTier/Handler.cs [hunk 6]
```diff
@@ -22,6 +23,10 @@ public Handler(ITiersRepository tierRepository, ILogger<Handler> logger, IEventB
 
     public async Task<CreateTierResponse> Handle(CreateTierCommand request, CancellationToken cancellationToken)
     {
+        var tierExists = await _tierRepository.ExistsWithName(request.Name, cancellationToken);
+        if (tierExists)
+            throw new OperationFailedException(ApplicationErrors.Devices.TierNameAlreadyExists());
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Application/Tiers/Commands/CreateTier/Handler.cs:28]: Can you throw an ApplicationException please? I don't remember why I even introduced the OperationFailedException. ;-)

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Application/Tiers/Commands/CreateTier/Handler.cs (hunk 6 above)
Comment (JLSRKonk) [reply Modules/Devices/src/Devices.Application/Tiers/Commands/CreateTier/Handler.cs:28]: Sure. I followed the CreateClients handler for this validation, and the handler there uses this same OperationFailedException. Should I also change it there?

Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Tiers/Commands/CreateTier/Handler.cs:28]: That would be great, thank you

**Review by tnotheis — APPROVED** [review]

**Pull request merged**